pkg/controller/master-controller-manager/constraint-template-controller/controller.go

```diff
@@ -0,0 +1,310 @@
+	seedClientProviders     map[string]*SeedClientProvider
+	userClusterClients      map[string]ctrlruntimeclient.Client
+}
+
+type SeedClientProvider struct {
+	seedClient                ctrlruntimeclient.Client
+	userClusterClientProvider UserClusterClientProvider
+}
+
+func Add(ctx context.Context,
+	mgr manager.Manager,
+	seedManagers map[string]manager.Manager,
+	log *zap.SugaredLogger,
+	workerName string,
+	numWorkers int) error {
+
+	workerSelector, err := workerlabel.LabelSelector(workerName)
+	if err != nil {
+		return fmt.Errorf("failed to build worker-name selector: %v", err)
+	}
+
+	reconciler := &reconciler{
+		ctx:                     ctx,
+		log:                     log.Named(ControllerName),
+		workerNameLabelSelector: workerSelector,
+		masterClient:            mgr.GetClient(),
+		seedClientProviders:     map[string]*SeedClientProvider{},
+		userClusterClients:      map[string]ctrlruntimeclient.Client{},
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: reconciler, MaxConcurrentReconciles: numWorkers})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %v", err)
+	}
+
+	for seedName, seedManager := range seedManagers {
+		seedClient := seedManager.GetClient()
+
+		userClusterClientProvider, err := getUserClusterClientProvider(seedManager)
+		if err != nil {
+			return fmt.Errorf("failed to get cluster client provider for seed %s: %w", seedName, err)
+		}
+		reconciler.seedClientProviders[seedName] = &SeedClientProvider{
+			seedClient:                seedClient,
+			userClusterClientProvider: userClusterClientProvider,
+		}
+
+		clusterSource := &source.Kind{Type: &kubermaticv1.Cluster{}}
+		if err := clusterSource.InjectCache(mgr.GetCache()); err != nil {
+			return fmt.Errorf("failed to inject cache into clusterSource for seed %s: %v", seedName, err)
+		}
+		if err := c.Watch(
+			clusterSource,
+			enqueueAllConstraintTemplates(reconciler.masterClient, reconciler.log),
+			workerlabel.Predicates(workerName),
+		); err != nil {
+			return fmt.Errorf("failed to establish watch for clusters in seed %s: %v", seedName, err)
+		}
+	}
+
+	if err := c.Watch(
+		&source.Kind{Type: &kubermaticv1.ConstraintTemplate{}},
+		&handler.EnqueueRequestForObject{},
+	); err != nil {
+		return fmt.Errorf("failed to create watch for constraintTemplates: %v", err)
+	}
+
+	return nil
+}
+
+// Reconcile reconciles the kubermatic constraint template on the master cluster to all seed cluster user clusters
+// which have opa integration enabled
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("ConstraintTemplate", request)
+	log.Debug("Reconciling")
+
+	constraintTemplate := &kubermaticv1.ConstraintTemplate{}
+	if err := r.masterClient.Get(r.ctx, request.NamespacedName, constraintTemplate); err != nil {
+		if kerrors.IsNotFound(err) {
+			log.Debugw("constraint template not found, returning", "constraint-name", constraintTemplate.Name)
```

Comment:
If it's not found, then constraintTemplate will be empty. Also, the name has already been assigned to the log in the `r.log.With()` line, no need to add it here again.